### PR TITLE
On GetLegendGraphics take the scale context from map parameters

### PR DESCRIPTION
**Pull request by signedav**
Taken from the already existing possibility to give with `GetLegendGraphics` request parameters `BBOX`, `WIDTH` and `HEIGHT` (used to only show legend entries from layers that have features on the defined area) it takes the scale to display the symbols with size depending from map units in the corrects size.

This is the minimum implementation to avoid to have always the defined max. size for the symbols or even worse to receive an error `createImage: image could not be created, check for out of memory conditions.` when no max. size defined. It would make sense to have a default value for the legend symbol size if there is not a given value for `BBOX`, `WIDTH` and `HEIGHT`...

Fixes #21309

### CODE REVIEW

File: src/core/qgslegendsettings.h [hunk 1]
```diff
@@ -178,6 +178,18 @@ class CORE_EXPORT QgsLegendSettings
      */
     void setMapScale( double scale ) { mMapScale = scale; }
 
+    /**
+     * Returns the factor of map units per pixel for symbols with size given in map units calculated by mDpi and mMmPerMapUnit
```
Comment (m-kuhn) [inline src/core/qgslegendsettings.h]: Better don't use private symbols in public API docs.

I.e. just write "dpi" and "mmPerMapUnit".

**Review by nyalldawson — COMMENTED** [review]
Don't forget a unit test too!

File: src/core/qgslegendsettings.h [hunk 2]
```diff
@@ -178,6 +178,18 @@ class CORE_EXPORT QgsLegendSettings
      */
     void setMapScale( double scale ) { mMapScale = scale; }
 
+    /**
+     * Returns the factor of map units per pixel for symbols with size given in map units calculated by dpi and mmPerMapUnit
+     * \see setMapUnitsPerPixel()
```
Comment (pblottiere) [inline src/core/qgslegendsettings.h]: Can you add `\since` too please?

File: src/core/qgslegendsettings.h [hunk 3]
```diff
@@ -178,6 +178,18 @@ class CORE_EXPORT QgsLegendSettings
...
+     */
+    double mapUnitsPerPixel() const;
+
+    /**
+     * Sets the the mmPerMapUnit calculated by \a mapUnitsPerPixel mostly taken from the map settings.
+     * \see mapUnitsPerPixel()
```
Comment (pblottiere) [inline src/core/qgslegendsettings.h]: `\since`

**Comment by signedav** [comment]
The tests will fail. Because the image size is not the same:

**Comment by nyalldawson** [comment]
> I created the test images by manual curl requests. Why do they differ?

Probably because of font rendering differences resulting in slightly different font metrics. The render checker has a setting setSizeTolerance to allow for this situation.

**Comment by signedav** [comment]
Thanks. I see it's not only the size - the default general tolerance of 100 is not enough to count it as a match when I compare them with the images generated over the  curl command. So I'm using now the images generated by the tests itself. This would be pointless in case they would visually differ from the curl-created ones, but they don't.
Or do you think it's risky? 
How did you create the images you are comparing to @elpaso ?

**Comment by pblottiere** [comment]
Hi @signedav,

> So I'm using now the images generated by the tests itself.

AFAIAC, that's exactly how I do it.

**Comment by signedav** [comment]
Thanks @pblottiere

**Comment by pblottiere** [comment]
Hi @signedav,

I think you should use bigger font in bold. It'll reduce the effect of rendering differences. 

Then you should just take the image generated by travis (available on CDash). 

Let me know if I can help.

**Comment by signedav** [comment]
Thanks @pblottiere 
I did that in my last commit (first I took wrong ones). But now another test is failing. Don't know why. I will check it tomorrow...

**Comment by signedav** [comment]
Thanks for restarting tests to whoever it was :-) So thats it.
I will provide the backports for 3.6 and 3.4 today.

File: src/core/qgslegendsettings.h [hunk 4]
```diff
@@ -178,6 +178,20 @@ class CORE_EXPORT QgsLegendSettings
      */
     void setMapScale( double scale ) { mMapScale = scale; }
 
+    /**
+     * Returns the factor of map units per pixel for symbols with size given in map units calculated by dpi and mmPerMapUnit
+     * \see setMapUnitsPerPixel()
+     * \since QGIS 3.6
```
Comment (pblottiere) [inline src/core/qgslegendsettings.h]: Last minute review: `QGIS 3.8`

File: src/core/qgslegendsettings.h [hunk 5]
```diff
@@ -178,6 +178,20 @@ class CORE_EXPORT QgsLegendSettings
...
+     */
+    double mapUnitsPerPixel() const;
+
+    /**
+     * Sets the the mmPerMapUnit calculated by \a mapUnitsPerPixel mostly taken from the map settings.
+     * \see mapUnitsPerPixel()
+     * \since QGIS 3.6
```
Comment (pblottiere) [inline src/core/qgslegendsettings.h]: `QGIS 3.8`

**Comment by pblottiere** [comment]
Hi @signedav,

Once the documentation updated, I'll merge your PR. Thanks for your work!

File: src/core/qgslegendsettings.h (hunk 4 above)
Comment (signedav) [reply src/core/qgslegendsettings.h]: Of course. I was optimistic on Friday :-)

Comment (signedav) [reply src/core/qgslegendsettings.h]: But I'll do a backport to 3.6 and 3.4, right?

Comment (pblottiere) [reply src/core/qgslegendsettings.h]: > But I'll do a backport to 3.6 and 3.4, right?

It would be great yes :).

**Comment by daniviga** [comment]
> when no max. size defined. It would make sense to have a default value for the legend symbol size if there is not a given value for `BBOX`, `WIDTH` and `HEIGHT`...

Why don't relying on `SYMBOLWIDTH` and `SYMBOLHEIGT` in such case?

**Pull request merged**